packages/core/src/internal/proc.js

```diff
@@ -534,73 +534,63 @@ export default function proc(env, iterator, parentContext, parentEffectId, meta,
     // Fork effects are non cancellables
   }
 
-  function runJoinEffect(t, cb) {
-    if (t.isRunning()) {
+  function runJoinEffect(taskOrTasks, cb) {
+    if (is.array(taskOrTasks)) {
+      if (taskOrTasks.length === 0) {
+        cb([])
+        return
+      }
+
+      const childCallbacks = createAllStyleChildCallbacks(taskOrTasks, cb)
+      taskOrTasks.forEach((t, i) => {
+        joinSingleTask(t, childCallbacks[i])
+      })
+    } else {
+      joinSingleTask(taskOrTasks, cb)
+    }
+  }
+
+  function joinSingleTask(taskToJoin, cb) {
+    if (taskToJoin.isRunning()) {
       const joiner = { task, cb }
-      cb.cancel = () => remove(t.joiners, joiner)
-      t.joiners.push(joiner)
+      cb.cancel = () => remove(taskToJoin.joiners, joiner)
+      taskToJoin.joiners.push(joiner)
     } else {
-      t.isAborted() ? cb(t.error(), true) : cb(t.result())
+      if (taskToJoin.isAborted()) {
+        cb(taskToJoin.error(), true)
+      } else {
+        cb(taskToJoin.result())
+      }
     }
   }
 
-  function runCancelEffect(taskToCancel, cb) {
-    if (taskToCancel === SELF_CANCELLATION) {
-      taskToCancel = task
+  function runCancelEffect(taskOrTasks, cb) {
+    if (taskOrTasks === SELF_CANCELLATION) {
+      cancelSingleTask(task)
+    } else if (is.array(taskOrTasks)) {
+      taskOrTasks.forEach(cancelSingleTask)
+    } else {
+      cancelSingleTask(taskOrTasks)
     }
+    cb()
```

Comment:
this got me thinking - what happens when we cancel a task? it calls `it.return()`, right? what happens if `finally` block throws? is that case handled? 🤔

Reply:
*cancelled* or *aborted* are two mutual-exclusive state for a task. So if we cancel a task (task's state becomes *cancelled*), all the subsequent errors from the task iterator should not bubble to the parent (or else task's state becomes *aborted*).

Code for handing error during cancellation is [here](https://github.com/redux-saga/redux-saga/blob/29fbd1b50d7128d3b9e4ff4105ae3b84ed2a2905/packages/core/src/internal/proc.js#L274-L276) as follows:
```js
// function next(arg, isErr)
try {
  // other code
} catch (error) {
  if (mainTask._isCancelled) {
    env.logError(error) // currently, we simply log the error if the task is already cancelled
  }
  // maybe we could add `else` to wrap the following two lines of code
  mainTask._isRunning = false
  mainTask.cont(error, true) // if `mainTask._isCancelled` true, here `mainTask.cont` is noop
}
```

Reply:
so it's handled - thanks for checking!